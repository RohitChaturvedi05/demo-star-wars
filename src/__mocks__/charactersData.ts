import {
    CharacterDetails,
    Characters,
    Film,
    Starship,
} from '../models/Character';

export const mockCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
    {
        uid: '2',
        id: '2',
        name: 'Leia Organa',
        gender: 'female',
        skin_color: 'light',
        hair_color: 'brown',
        height: '150',
        eye_color: 'brown',
        mass: '49',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/2/',
        homeworld: '2',
    },
];

export const mockFavoriteCharacters: Characters = [
    {
        uid: '1',
        id: '1',
        name: 'Luke Skywalker',
        gender: 'male',
        skin_color: 'fair',
        hair_color: 'blond',
        height: '172',
        eye_color: 'blue',
        mass: '77',
        birth_year: '19BBY',
        homeworldUrl: 'http://swapi.dev/api/planets/1/',
        homeworld: '1',
    },
];

export const mockPlanets = {
    1: {
        id: '1',
        name: 'Tatooine',
        diameter: '10465',
        gravity: '1',
        population: '200000',
    },
    2: {
        id: '2',
        name: 'Alderaan',
        diameter: '12500',
        gravity: '1',
        population: '2000000000',
    },
};

export const mockCharacterDetails: CharacterDetails = {
    uid: '1',
    id: '1',
    name: 'Luke Skywalker',
    gender: 'male',
    skin_color: 'fair',
    hair_color: 'blond',
    height: '172',
    eye_color: 'blue',
    mass: '77',
    birth_year: '19BBY',
    homeworldUrl: 'http://swapi.dev/api/planets/1/',
    homeworld: '1',
    films: ['http://swapi.dev/api/film/1', 'http://swapi.dev/api/film/2'],
    starships: [
        'http://swapi.dev/api/starship/1',
        'http://swapi.dev/api/starship/2',
    ],
};

export const mockFilms: Film[] = [
    {
        id: '1',
        title: 'A New Hope',
        director: 'George Lucas',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1977-05-25',
    },
    {
        id: '2',
        title: 'The Empire Strikes Back',
        director: 'Irvin Kershner',
        producer: 'Gary Kurtz, Rick McCallum',
        release_date: '1980-05-17',
    },
];

export const mockStarship: Starship[] = [
    {
        name: 'Millennium Falcon',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Corellian Engineering Corporation',
    },
    {
        name: 'X-wing',
        cargo_capacity: '100000',
        passengers: '6',
        max_atmosphering_speed: '1050',
        manufacturer: 'Incom Corporation',
    },
];
