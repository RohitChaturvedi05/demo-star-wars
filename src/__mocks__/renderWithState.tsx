import { render } from '@testing-library/react';
import React from 'react';
import { StateProvider } from '../state';

export const renderWithState = (Component: React.ReactNode) => {
    return render(<StateProvider>{Component}</StateProvider>);
};
