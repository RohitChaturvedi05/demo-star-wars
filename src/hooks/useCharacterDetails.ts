import { useCallback, useMemo, useState } from 'react';
import { Gender } from '../components/Gender';
import { Character, Film, Starship } from '../models/Character';
import { getCharacterById } from '../services/getCharacterById';
import { getFilm } from '../services/getFilms';
import { getPlanetsByIds } from '../services/getPlanets';
import { getStarShip } from '../services/getStarship';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharacterDetails = () => {
    const [message, setMessage] = useState('');
    const { state, dispatch } = useStateContext();
    const { planets } = state;
    const [character, setCharacter] = useState<Character>();
    const [films, setFilms] = useState<Film[]>([]);
    const [starShips, setStarShips] = useState<Starship[]>([]);

    const getCharacterDetails = useCallback(
        async (id: string) => {
            const characterData = await getCharacterById(id);
            if (characterData) {
                const planetId = parseInt(characterData?.homeworld ?? '');
                if (planets[planetId] === undefined) {
                    const data = await getPlanetsByIds([planetId]);
                    dispatch({ type: ActionType.SET_PLANETS, payload: data });
                }
                const filePromise = characterData?.films.map(getFilm);
                const starShipPromise =
                    characterData?.starships.map(getStarShip);

                const allFilms = await Promise.all(filePromise);
                const allStarShips = await Promise.all(starShipPromise);

                setFilms(allFilms.filter((film) => !!film) as Film[]);
                setStarShips(
                    allStarShips.filter((starShip) => !!starShip) as Starship[]
                );
                setCharacter(characterData);
            }
        },
        [dispatch, planets]
    );

    const onAddToFavorites = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const filmsNames = useMemo(
        () => films.map((film) => ({ name: film.title })),
        [films]
    );
    const starShipsNames = useMemo(
        () => starShips.map((starShip) => ({ name: starShip.name })),
        [starShips]
    );

    const isFavorite = useMemo(
        () => character && state.favCharacters.includes(character.id),
        [character, state.favCharacters]
    );

    const onGenderChange = (value: Gender, id: string) => {
        dispatch({ type: ActionType.SET_GENDER, payload: { value, id } });
        setTimeout(() => {
            setMessage('Gender Saved !!');
        }, 2000);
    };

    return {
        message,
        films,
        starShips,
        planets,
        character,
        filmsNames,
        isFavorite,
        starShipsNames,
        onGenderChange,
        onAddToFavorites,
        getCharacterDetails,
    };
};
