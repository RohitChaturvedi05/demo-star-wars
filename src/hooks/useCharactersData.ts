import { useCallback, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { getCharacters } from '../services/getCharacters';
import { getPlanetsByIds } from '../services/getPlanets';
import { getCharactersSearch } from '../services/searchByCharacterName';
import { extractPlanetId } from '../services/utils/extractPlanetId';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';
import { useDebounce } from './useDebounce';

export const useCharactersData = () => {
    const [queryTerm, setQueryTerm] = useState('');
    const debounceQuery = useDebounce(queryTerm, 300);

    const [isLoading, setLoading] = useState(false);
    const { state, dispatch } = useStateContext();
    const { planets } = state;

    const navigate = useNavigate();

    const getCharactersData = useCallback(
        async (page: number) => {
            setLoading(true);
            const { characters, total_pages } = await getCharacters({
                page,
            });

            // get planets ids from characters
            const planetsId = characters
                .map((item) => extractPlanetId(item.homeworldUrl))
                .filter((id): id is number => id !== null && id !== undefined)
                .filter((id) => planets[id] === undefined);

            // Update planets state if needed
            if (planetsId.length > 0) {
                const planetsData = await getPlanetsByIds(
                    Array.from(new Set(planetsId))
                );
                dispatch({
                    type: ActionType.SET_PLANETS,
                    payload: planetsData,
                });
            }
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: total_pages });
            dispatch({
                type: ActionType.SET_CURRENT_PAGE,
                payload: page,
            });

            setLoading(false);
        },
        [dispatch, planets]
    );

    const searchCharactersByName = useCallback(
        async (name: string) => {
            setLoading(true);
            const { characters } = await getCharactersSearch({
                page: 1,
                name,
            });
            dispatch({ type: ActionType.SET_CHARACTERS, payload: characters });
            dispatch({ type: ActionType.SET_TOTAL_PAGE, payload: 0 });
            setLoading(false);
        },
        [dispatch]
    );

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    const onPageChange = async (page: number) => {
        dispatch({
            type: ActionType.SET_CURRENT_PAGE,
            payload: page,
        });
        await getCharactersData(page);
    };

    const onSearch = async (searchTerm: string) => {
        setQueryTerm(searchTerm);
        setLoading(true);
        if (searchTerm.length === 0) {
            await onPageChange(1);
        }
    };

    const initializeState = useCallback(
        () => getCharactersData(1),
        [getCharactersData]
    );

    const favorites = useMemo(() => {
        return state.characters.filter(({ uid }) =>
            state.favCharacters.includes(uid)
        );
    }, [state.characters, state.favCharacters]);

    return {
        isLoading,
        favorites,
        queryTerm,
        debounceQuery,
        characters: state.characters,
        planets: state.planets,
        favCharacters: state.favCharacters,
        currentPage: state.currentPage,
        totalPage: state.totalPage,
        initializeState,
        onSearch,
        searchCharactersByName,
        onFavoriteToggle,
        onDetailClick,
        getCharactersData,
        onPageChange,
    };
};
