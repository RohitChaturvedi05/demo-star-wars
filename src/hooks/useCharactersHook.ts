import { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { APP_PATH } from '../routes/constants';
import { useStateContext } from '../state';
import { ActionType } from '../state/actions';

export const useCharactersHook = () => {
    const { dispatch } = useStateContext();
    const navigate = useNavigate();

    const onFavoriteToggle = useCallback(
        (id: string) => {
            dispatch({
                type: ActionType.SET_FAV_CHARACTERS,
                payload: id,
            });
        },
        [dispatch]
    );

    const onDetailClick = useCallback(
        (id: string) => {
            navigate(`${APP_PATH.DETAILS}/${id}`);
        },
        [navigate]
    );

    return {
        onFavoriteToggle,
        onDetailClick,
    };
};
