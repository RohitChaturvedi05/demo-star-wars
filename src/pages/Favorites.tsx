import { Toolbar, Typography } from '@mui/material';
import { styled } from 'styled-components';
import { CharacterList } from '../components/CharacterList';
import { useCharactersData } from '../hooks/useCharactersData';

const PageContainer = styled.div`
    padding: 24px;
    width: 100%;
`;

export const Favorites = () => {
    const {
        planets,
        favorites,
        favCharacters,
        onFavoriteToggle,
        onDetailClick,
    } = useCharactersData();

    return (
        <>
            <Toolbar />
            <PageContainer>
                <Typography variant="h4" gutterBottom>
                    Star Wars Characters
                </Typography>
                <CharacterList
                    planets={planets}
                    characters={favorites}
                    favCharacters={favCharacters}
                    showPagination={false}
                    onFavoriteToggle={onFavoriteToggle}
                    onDetailClick={onDetailClick}
                />
            </PageContainer>
        </>
    );
};

export default Favorites;
