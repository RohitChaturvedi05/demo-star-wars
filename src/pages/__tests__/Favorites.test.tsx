import { fireEvent, screen } from '@testing-library/react';
import {
    mockFavoriteCharacters,
    mockPlanets,
} from '../../__mocks__/charactersData';
import { renderWithState } from '../../__mocks__/renderWithState';
import { useCharactersData } from '../../hooks/useCharactersData';
import Favorites from '../Favorites';

// Acceptance Criteria - Favourites view
// o display all characters that have been added to the favourites list(name, height, gender & home planet)
// o provide the ability to remove characters from the list

jest.mock('react-router-dom', () => ({
    useNavigate: jest.fn(),
    useParams: jest.fn().mockReturnValue({ id: '1' }),
}));
jest.mock('../../hooks/useCharactersData');

const MockUseCharactersData = {
    isLoading: false,
    characters: [],
    planets: mockPlanets,
    currentPage: 1,
    totalPage: 3,
    favCharacters: ['1'],
    favorites: mockFavoriteCharacters,
    searchCharactersByName: jest.fn(),
    onPageChange: jest.fn(),
    getCharactersData: jest.fn(),
    onDetailClick: jest.fn(),
    onFavoriteToggle: jest.fn(),
};

describe('pages/Favorites', () => {
    beforeEach(() => {
        (useCharactersData as jest.Mock).mockReturnValue(MockUseCharactersData);
    });

    it('should render without crashing', () => {
        const { container } = renderWithState(<Favorites />);
        expect(container).toMatchSnapshot();
    });
    it('should render list of characters with name, gender and home planet', () => {
        renderWithState(<Favorites />);
        expect(screen.getByText('Luke Skywalker')).toBeInTheDocument();
        expect(screen.getByText('male')).toBeInTheDocument();
        expect(screen.getByText('Tatooine')).toBeInTheDocument();
    });
    it('should add/remove from favorites', () => {
        renderWithState(<Favorites />);
        const favBtn = screen.getByTestId('btn-fav-1');
        fireEvent.click(favBtn);
        expect(MockUseCharactersData.onFavoriteToggle).toHaveBeenCalledWith(
            '1'
        );
    });
});
