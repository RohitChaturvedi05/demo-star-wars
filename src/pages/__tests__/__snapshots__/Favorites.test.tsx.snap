// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Favorites should render without crashing 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
  />
  <div
    class="sc-FEMpB jnBMmq"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
    >
      Star Wars Characters
    </h4>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
      data-testid="character-list"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <table
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
        data-testid="character-list$table"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          data-testid="character-list$head"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Gender
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            >
              Home planet
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          data-testid="character-list$body"
        >
          <tr
            class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-fav-1"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="StarIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                  />
                </svg>
              </button>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Luke Skywalker
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              male
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              Tatooine
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                data-testid="btn-details-1"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="ListAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5zm1.1-2H3.9c-.5 0-.9.4-.9.9v16.2c0 .4.4.9.9.9h16.2c.4 0 .9-.5.9-.9V3.9c0-.5-.5-.9-.9-.9M11 7h6v2h-6zm0 4h6v2h-6zm0 4h6v2h-6zM7 7h2v2H7zm0 4h2v2H7zm0 4h2v2H7z"
                  />
                </svg>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
